# Ford GoBike System Data Exploration
## Oladimeji Olaniyan 

## Introduction

This dataset includes information about individual rides made in a bike-sharing system covering the greater San Francisco Bay area for the month of February, 2019.

## Preliminary Wrangling


In [19]:
# import all packages and set plots to be embedded inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb
import datetime as dt

%matplotlib inline

### Gathering

In [3]:
# loading the dataset into a pandas dataframe
df = pd.read_csv(r'C:\Users\PC\Documents\GitHub\ALX-T\Project 03 - Data Visualization\Datasets\201902-fordgobike-tripdata.csv')

### Assessing

In [4]:
# we will be focusing our attention on the programmatic assessment of the dataset.
# writing a simple function to return high level overview of data shape and composition
def info(df):
    print(df.shape),
    print('*' *40),
    print(df.info())
    print('*' *40)
    print(df.head())

info(df)

(183412, 16)
****************************************
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 183412 entries, 0 to 183411
Data columns (total 16 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   duration_sec             183412 non-null  int64  
 1   start_time               183412 non-null  object 
 2   end_time                 183412 non-null  object 
 3   start_station_id         183215 non-null  float64
 4   start_station_name       183215 non-null  object 
 5   start_station_latitude   183412 non-null  float64
 6   start_station_longitude  183412 non-null  float64
 7   end_station_id           183215 non-null  float64
 8   end_station_name         183215 non-null  object 
 9   end_station_latitude     183412 non-null  float64
 10  end_station_longitude    183412 non-null  float64
 11  bike_id                  183412 non-null  int64  
 12  user_type                183412 non-null  object 
 13  membe

In [5]:
# this returns summary statistics on the numerical variables in the dataset.
df.describe()

duration_sec  start_station_id  start_station_latitude  \
count  183412.000000     183215.000000           183412.000000   
mean      726.078435        138.590427               37.771223   
std      1794.389780        111.778864                0.099581   
min        61.000000          3.000000               37.317298   
25%       325.000000         47.000000               37.770083   
50%       514.000000        104.000000               37.780760   
75%       796.000000        239.000000               37.797280   
max     85444.000000        398.000000               37.880222   

       start_station_longitude  end_station_id  end_station_latitude  \
count            183412.000000   183215.000000         183412.000000   
mean               -122.352664      136.249123             37.771427   
std                   0.117097      111.515131              0.099490   
min                -122.453704        3.000000             37.317298   
25%                -122.412408       44.000000             37.770407   
50%                -122.398285      100.000000             37.781010   
75%                -122.286533      235.000000             37.797320   
max                -121.874119      398.000000             37.880222   

       end_station_longitude        bike_id  member_birth_year  
count          183412.000000  183412.000000      175147.000000  
mean             -122.352250    4472.906375        1984.806437  
std                 0.116673    1664.383394          10.116689  
min              -122.453704      11.000000        1878.000000  
25%              -122.411726    3777.000000        1980.000000  
50%              -122.398279    4958.000000        1987.000000  
75%              -122.288045    5502.000000        1992.000000  
max              -121.874119    6645.000000        2001.000000

In [6]:
# checking all columns for unique value count
print(df.nunique())

duration_sec                 4752
start_time                 183401
end_time                   183397
start_station_id              329
start_station_name            329
start_station_latitude        334
start_station_longitude       335
end_station_id                329
end_station_name              329
end_station_latitude          335
end_station_longitude         335
bike_id                      4646
user_type                       2
member_birth_year              75
member_gender                   3
bike_share_for_all_trip         2
dtype: int64


### Issues

#### Quality
1. Null Values: start_station_id, start_station_name, end_station_id, end_station_name, member_birth_year, member_gender

2. Erroneous Datatype: start_time, end_time, start_station_id, end_station_id, user_type, member_gender, bike_share_for_all_trip.


#### Tidiness

3. We need to separate the start_time and end_time columns into time_of_day, day and month columns. 

### Cleaning 

In [7]:
# creating a copy of the dataset before cleaning
df_clean = df.copy()

df_clean.head()

duration_sec                start_time                  end_time  \
0         52185  2019-02-28 17:32:10.1450  2019-03-01 08:01:55.9750   
1         42521  2019-02-28 18:53:21.7890  2019-03-01 06:42:03.0560   
2         61854  2019-02-28 12:13:13.2180  2019-03-01 05:24:08.1460   
3         36490  2019-02-28 17:54:26.0100  2019-03-01 04:02:36.8420   
4          1585  2019-02-28 23:54:18.5490  2019-03-01 00:20:44.0740   

   start_station_id                                start_station_name  \
0              21.0  Montgomery St BART Station (Market St at 2nd St)   
1              23.0                     The Embarcadero at Steuart St   
2              86.0                           Market St at Dolores St   
3             375.0                           Grove St at Masonic Ave   
4               7.0                               Frank H Ogawa Plaza   

   start_station_latitude  start_station_longitude  end_station_id  \
0               37.789625              -122.400811            13.0   
1               37.791464              -122.391034            81.0   
2               37.769305              -122.426826             3.0   
3               37.774836              -122.446546            70.0   
4               37.804562              -122.271738           222.0   

                               end_station_name  end_station_latitude  \
0                Commercial St at Montgomery St             37.794231   
1                            Berry St at 4th St             37.775880   
2  Powell St BART Station (Market St at 4th St)             37.786375   
3                        Central Ave at Fell St             37.773311   
4                         10th Ave at E 15th St             37.792714   

   end_station_longitude  bike_id   user_type  member_birth_year  \
0            -122.402923     4902    Customer             1984.0   
1            -122.393170     2535    Customer                NaN   
2            -122.404904     5905    Customer             1972.0   
3            -122.444293     6638  Subscriber             1989.0   
4            -122.248780     4898  Subscriber             1974.0   

  member_gender bike_share_for_all_trip  
0          Male                      No  
1           NaN                      No  
2          Male                      No  
3         Other                      No  
4          Male                     Yes

#### Issue 1: Null Values: start_station_id, start_station_name, end_station_id, end_station_name, member_birth_year, member_gender

#### Define

Dataset contains null values which need to be dropped.

#### Code

In [9]:
# test
df_clean.isnull().any().sum()

6

In [10]:
df_clean.dropna(inplace=True)

#### Test

In [12]:
df_clean.isnull().any().sum()

0

#### Issue 2: Erroneous Datatype: start_time, end_time, start_station_id, end_station_id, user_type, member_gender, bike_share_for_all_trip.

#### Define

Columns in the wrong data type will be converted to the appropriate data type.

#### Code

In [13]:
# test
print(df_clean.start_time.dtypes)
print(df_clean.end_time.dtypes)
print(df_clean.start_station_id.dtypes)
print(df_clean.end_station_id.dtypes)
print(df_clean.member_gender.dtypes)
print(df_clean.user_type.dtypes)
print(df_clean.bike_share_for_all_trip.dtypes)

object
object
float64
float64
object
object
object


In [14]:
# we will use the pd.to_datetime function to convert the start_date and end_date columns to date data type
df_clean['start_time'] = pd.to_datetime(df_clean['start_time'])
df_clean['end_time'] = pd.to_datetime(df_clean['end_time'])

# converting the rest of the columns with erroneous datatypes
df_clean['start_station_id'] = df_clean.start_station_id.astype(int)
df_clean['end_station_id'] = df_clean.end_station_id.astype(int)
df_clean['member_gender'] = df_clean.member_gender.astype('category')
df_clean['user_type'] =df_clean.user_type.astype('category')
df_clean['bike_share_for_all_trip'] = df_clean.bike_share_for_all_trip.astype('category')

#### Test

In [15]:
print(df_clean.start_time.dtypes)
print(df_clean.end_time.dtypes)
print(df_clean.start_station_id.dtypes)
print(df_clean.end_station_id.dtypes)
print(df_clean.member_gender.dtypes)
print(df_clean.user_type.dtypes)
print(df_clean.bike_share_for_all_trip.dtypes)

datetime64[ns]
datetime64[ns]
int32
int32
category
category
category


#### Issue 3: We need to separate the start_time and end_time columns into time_of_day, day_of_week and month columns.

#### Define

In order to effectively explore certain metrics from the dataset during exploratory analysis, we will need to extract time of day, day of week and month from the start_date and end_date columns.

#### Code

In [20]:
# using the python datetime library, we will extract the needed info into new columns
# start_time
df_clean['start_time_of_day'] = df_clean['start_time'].dt.hour
df_clean['start_day_of_week'] = df_clean['start_time'].dt.day_name()
df_clean['start_month'] = df_clean['start_time'].dt.month_name()

#end_time
df_clean['end_time_of_day'] = df_clean['end_time'].dt.hour
df_clean['end_day_of_week'] = df_clean['end_time'].dt.day_name()
df_clean['end_month'] = df_clean['end_time'].dt.month_name()

#### Test

In [23]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 174952 entries, 0 to 183411
Data columns (total 22 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   duration_sec             174952 non-null  int64         
 1   start_time               174952 non-null  datetime64[ns]
 2   end_time                 174952 non-null  datetime64[ns]
 3   start_station_id         174952 non-null  int32         
 4   start_station_name       174952 non-null  object        
 5   start_station_latitude   174952 non-null  float64       
 6   start_station_longitude  174952 non-null  float64       
 7   end_station_id           174952 non-null  int32         
 8   end_station_name         174952 non-null  object        
 9   end_station_latitude     174952 non-null  float64       
 10  end_station_longitude    174952 non-null  float64       
 11  bike_id                  174952 non-null  int64         
 12  user_type       

### What is the structure of your dataset?
The dataset originally contained 16 columns and 183,412 rows that provide information on a bike-sharing system covering the greater San Francisco Bay area. However, after assessing and cleaning the dataset, it now contains 22 columns and 174,952 rows. The start_time and end_time columns were further split into 3 new column each which contain the time of day, day of week and month of the start and end times.

This dateset includes the bike ID, start and end time of each bike ride, with another column containing the duration of these bike rides in seconds.<br>
It also contains location information on the stations, such as the ID, name, latitude and longitude of the start and end stations.<br>
And lastly, the dataset also gives us information on the riders such as, the type of user ; whether they are a customer or a subscriber. The gender of the riders, as well as the birth year of the riders. 


### What is/are the main feature(s) of interest in your dataset?
What features in the dataset have the most effect on ride duration?

### What features in the dataset do you think will help support your investigation into your feature(s) of interest?
I expect the features that would have the most effect on ride duration in the dataset are what hour of the day the ride starts, and what day of the week the ride starts. 

Other features that may affect ride duration include the start stations of the ride, user type i.e whether they are a customer or a subscriber, and the gender of the rider.

## Univariate Exploration

> In this section, investigate distributions of individual variables. If
you see unusual points or outliers, take a deeper look to clean things up
and prepare yourself to look at relationships between variables.


> **Rubric Tip**: The project (Parts I alone) should have at least 15 visualizations distributed over univariate, bivariate, and multivariate plots to explore many relationships in the data set.  Use reasoning to justify the flow of the exploration.



>**Rubric Tip**: Use the "Question-Visualization-Observations" framework  throughout the exploration. This framework involves **asking a question from the data, creating a visualization to find answers, and then recording observations after each visualisation.** 




>**Rubric Tip**: Visualizations should depict the data appropriately so that the plots are easily interpretable. You should choose an appropriate plot type, data encodings, and formatting as needed. The formatting may include setting/adding the title, labels, legend, and comments. Also, do not overplot or incorrectly plot ordinal data.

### Discuss the distribution(s) of your variable(s) of interest. Were there any unusual points? Did you need to perform any transformations?

> Your answer here!

### Of the features you investigated, were there any unusual distributions? Did you perform any operations on the data to tidy, adjust, or change the form of the data? If so, why did you do this?

> Your answer here!

## Bivariate Exploration

> In this section, investigate relationships between pairs of variables in your
data. Make sure the variables that you cover here have been introduced in some
fashion in the previous section (univariate exploration).

### Talk about some of the relationships you observed in this part of the investigation. How did the feature(s) of interest vary with other features in the dataset?

> Your answer here!

### Did you observe any interesting relationships between the other features (not the main feature(s) of interest)?

> Your answer here!

## Multivariate Exploration

> Create plots of three or more variables to investigate your data even
further. Make sure that your investigations are justified, and follow from
your work in the previous sections.

### Talk about some of the relationships you observed in this part of the investigation. Were there features that strengthened each other in terms of looking at your feature(s) of interest?

> Your answer here!

### Were there any interesting or surprising interactions between features?

> Your answer here!

## Conclusions
>You can write a summary of the main findings and reflect on the steps taken during the data exploration.



> Remove all Tips mentioned above, before you convert this notebook to PDF/HTML


> At the end of your report, make sure that you export the notebook as an
html file from the `File > Download as... > HTML or PDF` menu. Make sure you keep
track of where the exported file goes, so you can put it in the same folder
as this notebook for project submission. Also, make sure you remove all of
the quote-formatted guide notes like this one before you finish your report!

